### ISSUE TITLE
Documentation about C++ exceptions may leads to errors

### ISSUE THREAD

**Issue by MrHalfman** [swig/swig#789]
Hello,

Just a little message about how C++ exceptions handling is described in the documentation especially when targeting JavaScript : the documentation is explaining us to use something like this : 
```
%exception {
  try {                                                     
    $action
  } catch (const std::runtime_error &e) {
    SWIG_exception(SWIG_RuntimeError, e.what());
  } catch (...) {
    SWIG_exception(SWIG_RuntimeError, "unknown exception");
  } 
} 
``` 

It's working for most of the situations but when an error is thrown in the constructor of an object, it makes the app crashing (by calling the destructor while the object wasn't created). I found a solution by adding a `goto fail;`after the calls to `SWIG_exception`which seems to work actually.

Maybe the documentation needs an update to deal with this situation :wink:.

Thanks for making this nice tool which saved a lot of my time, I hope this helps =).

**Comment by ojwb** [comment]
I think this is actually a bug in SWIG's Javascript backend - I can't see anywhere that actually clearly defines this behaviour, but for most languages `SWIG_exception()` seems to include `SWIG_fail;` (which typically expands to `goto fail;`) or some alternative way to return from the function it's used in - looking at the languages where it's fairly obvious from `grep`:

```
Lib/go/exception.i:#define SWIG_exception(code, msg) _swig_gopanic(msg)
Lib/octave/exception.i:  %define_as(SWIG_exception(code, msg), %block(%error(code, msg); SWIG_fail; ))
Lib/perl5/exception.i:  %define_as(SWIG_exception(code, msg), %block(%error(code, msg); SWIG_fail; ))
Lib/python/exception.i:  %define_as(SWIG_exception(code, msg), %block(%error(code, msg); SWIG_fail; ))
Lib/tcl/exception.i:  %define_as(SWIG_exception(code, msg), %block(%error(code, msg); return TCL_ERROR;))
```

Ruby's version ends up calling `rb_raise()` which apparently doesn't return.

PHP's version currently continues, but that appears to be a bug (reported at #240, PR #785 fixes this, but lacks a testcase).

@wsfulton Is that correct?

And if it is, why is there also `SWIG_exception_fail`?

**Comment by ojwb** [comment]
26bbc96d58b81293a2357ba3ba7c4419d87417dd added a new `swig_exception` testcase for this, which gives these warnings with Javascript because `SWIG_exception` doesn't return there:

```
checking javascript (node) testcase swig_exception
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_new_Square(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2216:75: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_new_Circle(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2080:75: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_new_Value(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:1664:74: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Square_perimeter(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2282:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Square_area(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2248:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Circle_perimeter(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2146:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Circle_area(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2112:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Shape_perimeter(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:1963:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Shape_area(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:1929:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
```

**Comment by ojwb** [comment]
I've pushed a candidate fix in #792 for CI testing.

**Comment by MrHalfman** [comment]
Thanks for the fix =).
